[inductor] Make {cudagraph_trees,decomposition,post_grad}.py pass follow_imports typechecking

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113609

I added explicit imports to `kernel/__init__.py` as mypy doesn't seem to
understand an empty `__init__.py`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

int3 committed:
[inductor] Make {cudagraph_trees,decomposition,post_grad}.py pass follow_imports typechecking

I added explicit imports to `kernel/__init__.py` as mypy doesn't seem to
understand an empty `__init__.py`.

[ghstack-poisoned]

int3 committed:
Update on "[inductor] Make {cudagraph_trees,decomposition,post_grad}.py pass follow_imports typechecking"


I added explicit imports to `kernel/__init__.py` as mypy doesn't seem to
understand an empty `__init__.py`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -714,6 +713,8 @@ def __call__(
         # generating a random torch.Tensor for benchmarking.
         input_gen_fns: Optional[Dict[int, Callable[[ir.Buffer], torch.Tensor]]] = None,
     ):
+        from .codegen.cuda.cuda_kernel import CUDATemplateCaller
```
The explicit imports in `kernel/__init__.py` above changed the order of module loading and exposed a circular import between `select_algorithm.py` and `cuda_kernel.py`.

int3 commented:
Most of the declarations added to `torch/_C/__init__.pyi.in` were to support references from `cudagraph_trees.py`.

eellison commented on torch/_inductor/decomposition.py:
```diff
@@ -348,7 +348,7 @@ def get_like_layout(
 ) -> torch.memory_format:
     # TODO: _to_copy tensor to stride permutation
-    if memory_format in (torch.preserve_format, None):
+    if memory_format is torch.preserve_format or memory_format is None:
```
just curious - why was this required

eellison commented on torch/_inductor/cudagraph_trees.py:
```diff
@@ -969,7 +969,7 @@ def reconstruct_outputs(self):
             self._initialize_cached_tensors()
 
-        outputs: List[torch.Tensor] = []
+        outputs: List[Any] = []
```
nit: `List[Union[int, None, torch.Tensor]]`

int3 replied:
IIRC this created a lot of nullcheck issues but let me try again

eellison reviewed (approved):
looks great!

int3 committed:
Update on "[inductor] Make {cudagraph_trees,decomposition,post_grad}.py pass follow_imports typechecking"


I added explicit imports to `kernel/__init__.py` as mypy doesn't seem to
understand an empty `__init__.py`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 committed:
Update on "[inductor] Make {cudagraph_trees,decomposition,post_grad}.py pass follow_imports typechecking"


I added explicit imports to `kernel/__init__.py` as mypy doesn't seem to
understand an empty `__init__.py`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
@pytorchbot merge
